crypto/sm2/sm2_crypt.c

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2017 Ribose Inc. All Rights Reserved.
+ * Ported from Ribose contributions from Botan.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/sm2.h>
+#include <openssl/evp.h>
+#include <openssl/asn1.h>
+#include <openssl/asn1t.h>
+#include <string.h>
+
+typedef struct SM2_Ciphertext_st SM2_Ciphertext;
+DECLARE_ASN1_FUNCTIONS(SM2_Ciphertext)
+
+struct SM2_Ciphertext_st {
+    BIGNUM *C1x;
+    BIGNUM *C1y;
+    ASN1_OCTET_STRING *C3;
+    ASN1_OCTET_STRING *C2;
+};
+
+ASN1_SEQUENCE(SM2_Ciphertext) =
+{
+ASN1_SIMPLE(SM2_Ciphertext, C1x, BIGNUM),
+        ASN1_SIMPLE(SM2_Ciphertext, C1y, BIGNUM),
+        ASN1_SIMPLE(SM2_Ciphertext, C3, ASN1_OCTET_STRING),
+        ASN1_SIMPLE(SM2_Ciphertext, C2,
+                        ASN1_OCTET_STRING),} ASN1_SEQUENCE_END(SM2_Ciphertext)
+IMPLEMENT_ASN1_FUNCTIONS(SM2_Ciphertext)
+
+static int EC_field_size(const EC_GROUP *group)
+{
+    /* Is there some simpler way to do this? */
+    BIGNUM *p = BN_new();
+    BIGNUM *a = BN_new();
+    BIGNUM *b = BN_new();
+    int field_size;
+
+    EC_GROUP_get_curve_GFp(group, p, a, b, NULL);
+    field_size = (BN_num_bits(p) + 7) / 8;
+
+    BN_free(p);
+    BN_free(a);
+    BN_free(b);
+
+    return field_size;
+}
+
+int SM2_ciphertext_size(const EC_KEY *key, const EVP_MD *digest, size_t msg_len)
+{
+    return 10 + 2 * EC_field_size(EC_KEY_get0_group(key)) +
+        EVP_MD_size(digest) + msg_len;
+}
+
+int SM2_encrypt(const EC_KEY *key,
+                const EVP_MD *digest,
+                const uint8_t *msg,
+                size_t msg_len, uint8_t *ciphertext_buf, size_t *ciphertext_len)
+{
+    int rc = 0;
+    BIGNUM *k = BN_new();
+    BIGNUM *x1 = BN_new();
+    BIGNUM *y1 = BN_new();
+    BIGNUM *x2 = BN_new();
+    BIGNUM *y2 = BN_new();
+
+    BN_CTX *ctx = BN_CTX_new();
+    EVP_MD_CTX *hash = EVP_MD_CTX_new();
+
+    const EC_GROUP *group = EC_KEY_get0_group(key);
+    const BIGNUM *order = EC_GROUP_get0_order(group);
+    EC_POINT *kG = EC_POINT_new(group);
+    EC_POINT *kP = EC_POINT_new(group);
+    const EC_POINT *P = EC_KEY_get0_public_key(key);
+    uint8_t msg_mask[msg_len];
```

Comment:
ANSI C requires a constant expression for the size.  You may need to allocate `msg_mask` on the heap